BaselineExactDependencies should filter for only jar artifacts

We use the jar analyzer, meaning it breaks when taking a dependency on a pom file for example

This manifests in 
```
Caused by: java.lang.IllegalArgumentException: Cannot accept visitor on URL: file:/Users/robertk/.gradle/caches/modules-2/files-2.1/org.apache.spark/spark-dist_2.11-hadoop-palantir/3.0.0-palantir.32/5413978ef8b201b92ba5b6514d8d9167a778d9e3/spark-dist_2.11-hadoop-palantir-3.0.0-palantir.32.pom
        at org.apache.maven.shared.dependency.analyzer.ClassFileVisitorUtils.accept(ClassFileVisitorUtils.java:75)
        at org.apache.maven.shared.dependency.analyzer.DefaultClassAnalyzer.analyze(DefaultClassAnalyzer.java:48)
        at com.palantir.baseline.plugins.BaselineExactDependencies$Indexes.lambda$populateIndexes$2(BaselineExactDependencies.java:106)
        at com.palantir.baseline.plugins.BaselineExactDependencies$Indexes.populateIndexes(BaselineExactDependencies.java:103)
        at com.palantir.baseline.tasks.CheckUnusedDependenciesTask.checkUnusedDependencies(CheckUnusedDependenciesTask.java:67)
```

--- a/changelog/@unreleased/pr-773.v2.yml
+++ b/changelog/@unreleased/pr-773.v2.yml
@@ -1,0 +1,5 @@
+type: fix
+fix:
+  description: Remove pom only dependencies from analysis in checkUnusedDependencies
+  links:
+  - https://github.com/palantir/gradle-baseline/pull/773

--- a/gradle-baseline-java/src/main/groovy/com/palantir/baseline/plugins/BaselineExactDependencies.java
+++ b/gradle-baseline-java/src/main/groovy/com/palantir/baseline/plugins/BaselineExactDependencies.java
@@ -17,6 +17,7 @@
 package com.palantir.baseline.plugins;
 
 import com.google.common.base.Preconditions;
+import com.google.common.collect.ImmutableSet;
 import com.palantir.baseline.tasks.CheckImplicitDependenciesTask;
 import com.palantir.baseline.tasks.CheckUnusedDependenciesTask;
 import java.io.File;
@@ -50,6 +51,7 @@ public final class BaselineExactDependencies implements Plugin<Project> {
     // All applications of this plugin share a single static 'Indexes' instance, because the classes
     // contained in a particular jar are immutable.
     public static final Indexes INDEXES = new Indexes();
+    public static final ImmutableSet<String> VALID_ARTIFACT_EXTENSIONS = ImmutableSet.of("jar", "");
 
     @Override
     public void apply(Project project) {
@@ -98,6 +100,7 @@ public final class BaselineExactDependencies implements Plugin<Project> {
         public void populateIndexes(Set<ResolvedDependency> declaredDependencies) {
             Set<ResolvedArtifact> allArtifacts = declaredDependencies.stream()
                     .flatMap(dependency -> dependency.getAllModuleArtifacts().stream())
+                    .filter(dependency -> VALID_ARTIFACT_EXTENSIONS.contains(dependency.getExtension()))
                     .collect(Collectors.toSet());
 
             allArtifacts.forEach(artifact -> {
@@ -111,10 +114,8 @@ public final class BaselineExactDependencies implements Plugin<Project> {
                 }
             });
 
-            declaredDependencies.stream().forEach(dependency -> {
-                Set<ResolvedArtifact> artifacts = dependency.getModuleArtifacts();
-                artifacts.forEach(artifact -> artifactsFromDependency.put(artifact, dependency));
-            });
+            declaredDependencies.forEach(dependency -> dependency.getModuleArtifacts()
+                    .forEach(artifact -> artifactsFromDependency.put(artifact, dependency)));
         }
 
         /** Given a class, what dependency brought it in. */

--- a/gradle-baseline-java/src/main/groovy/com/palantir/baseline/tasks/CheckUnusedDependenciesTask.java
+++ b/gradle-baseline-java/src/main/groovy/com/palantir/baseline/tasks/CheckUnusedDependenciesTask.java
@@ -74,6 +74,8 @@ public class CheckUnusedDependenciesTask extends DefaultTask {
                 .collect(Collectors.toSet());
         Set<ResolvedArtifact> declaredArtifacts = declaredDependencies.stream()
                 .flatMap(dependency -> dependency.getModuleArtifacts().stream())
+                .filter(dependency -> BaselineExactDependencies.VALID_ARTIFACT_EXTENSIONS
+                        .contains(dependency.getExtension()))
                 .collect(Collectors.toSet());
 
         List<ResolvedArtifact> declaredButUnused = Sets.difference(declaredArtifacts, necessaryArtifacts).stream()
@@ -93,6 +95,8 @@ public class CheckUnusedDependenciesTask extends DefaultTask {
                 ResolvedDependency dependency = BaselineExactDependencies.INDEXES
                         .artifactsFromDependency(resolvedArtifact);
                 Set<ResolvedArtifact> didYouMean = dependency.getAllModuleArtifacts().stream()
+                        .filter(artifact -> BaselineExactDependencies.VALID_ARTIFACT_EXTENSIONS
+                                .contains(artifact.getExtension()))
                         .flatMap(BaselineExactDependencies.INDEXES::classesFromArtifact)
                         .filter(referencedClasses()::contains)
                         .map(BaselineExactDependencies.INDEXES::classToDependency)

